### ISSUE TITLE
Pass variables inside fig.yml on runtime

### ISSUE THREAD

**Issue by spiddy** [docker/compose#495]
Is it possible to configure a fig.yml in such way that you can pass on execution time variables instead of hard-wiring them inside the fig.yml?

I think a generic variable injection during execution would be quite useful for various use cases.

e.g:

<pre>
jenkins:
  image: aespinosa/jenkins:latest
  ports:
    - "8080"
  hostname: ${HOSTNAME}
</pre>

HOSTNAME=ci fig up

That could inject the variable HOSTNAME inside the fig.yml during execution and execute a docker run with hostname `ci`.

ps. This is different than passing environment variables inside docker which is already supported (http://www.fig.sh/yml.html#environment)

**Comment by nathanpalmer** [comment]
+1 - This could be useful for our postgres image. Otherwise we aren't able to share the same configuration.

```
db:
  image: postgres:9.3
  volumes:
    - /$APP_NAME/postgresql/:/var/lib/postgresql/data/
  environment:
    - POSTGRES_USER=$CURRENT_USER
    - POSTGRES_PASSWORD=
  ports:
    - "5432:5432"
```

**Comment by JonLundy** [comment]
+1

**Comment by danggrianto** [comment]
+1

**Comment by NoumanSaleem** [comment]
:+1:

**Comment by relwell** [comment]
I have a few suggestions for how to implement this change:

* Fig should support two separate methods for accessing this data:
  * Environment variables, namespaced with FIG_VARIABLE_X
    * Example: If you're looking for HOSTNAME in fig.yml, the environment variable should be stored as FIG_VARIABLE_HOSTNAME
  * A separate YAML file
     * By default, we would search the root folder's variables.yml file
     * The user can specify the location of this file as an environment variable at FIG_VARIABLES_YML
     * The user can also pass this file via the command line for any command as --fig-variables=/foo/bar/variables.yml
* Create a class called FigConfig which wraps the all access of config data from fig.yml
  * This config class should have all the APIs available from the return value of yaml.safe_load to minimize changes to interactions throughout the library
   * This config class should include hooks for interpolating values first from templates.yaml, overwriteable by values set as environment variables
* Replace all existing calls to yaml.safe_load with instantiating a new instance of FigConfig with the provided filename. 

If everyone feels good about this proposal, I would be happy to get to work on it and issue a PR.

**Comment by mingfang** [comment]
+1
Can't wait!

**Comment by dnephin** [comment]
We already support environment variables in some fields without this name mangling. `${HOSTNAME}` would be used directly. I think this is a lot more intuitive than having to namespace them.

How would the values in this yaml file be represented?  Would it make more sense as a file with just one  `key=value` per line? Since this is more of a "nice to have" , I would probably wait on this.

> FigConfig

I like the idea of a class instead of a dict. It makes validation easier. You can centralize all config validation in that class (or before creating it), instead of scattering validation all over. I would not try to emulate the `dict` interface. I think you lose a lot of the benefit of using a class if you do that.


Another related request is to be able to provide default values for things. Something like: `${HOSTNAME:localhost}` is used in other files (tox.ini at least, possibly others), to provide defaults.

A change to support environment variables in most fields might be a good time to introduce this feature.


As a first past, I would suggest just adding extra support for environment variables in many fields using `os.environ()` is probably the easiest way to support this with minimal code changes.

**Comment by spiddy** [comment]
+1 on avoiding namespacing, since it is already provided as such in lot of
fields.
+1 also on providing default values, this is actually really cool,
otherwise by putting variables without defaults you miss the fanciness of
`fig up`

**Comment by relwell** [comment]
On the namespacing stuff, I think that's just a case of wanting to avoid interactions or dependencies between other systems on someone's machine. I'm happy to make namespacing environment variables a fully separate discussion, especially since it sounds like there's already support for some bare environment variables.

I think you're right that a bare INI-style format makes more sense here.  I initially wanted to avoid introducing "yet another" form of config management into this proposal.

So maybe the proposal should be something more like:

* Add access to os.environ() for all fig.yml fields
* Introduce a FigConfig class responsible for wrapping data access from fig.yml
* The FigConfig class should have hooks that can access data from the environment (provided the ${} format). The default value should be provided with a semicolon for the provided string.

We can de-scope the template file for now, since users could create a separate file with their environment variables to source from for the time being.

I'm happy to try to take this chunk of work, either in one PR or two. I'm leaning towards doing it in a single PR since we'd actually want to take os.environ() out of all currently exposed fields and have that access happen directly in the Config class.

**Comment by zakkie** [comment]
+1

**Comment by acofer** [comment]
+1

**Comment by marcellmars** [comment]
+1

**Comment by rejinka** [comment]
+1, especially @relwell's point of having a seperate YAML file

**Comment by ohord** [comment]
+1

**Comment by loa** [comment]
+1

Especially like this one:

```yaml
# fig.yml
app:
  image: django:${DJANGO_VERSION:latest}
  environment:
    - SECRET_KEY=${SECRET_KEY:ONLYFORTESTING}
```

**Comment by hadim** [comment]
+1

**Comment by NoumanSaleem** [comment]
fyi environment variables are capable of being referenced when declared without a value as the docs state 

>Environment variables with only a key are resolved to their values on the machine Fig is running on, which can be helpful for secret or host-specific values.

For the rest, try [envsubst](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) which can replace variable references using either `$var` or `${var}` syntax

```bash
envsubst < "fig.tpl" > "fig.yml"
fig up
```

**Comment by thomasdavis** [comment]
+1 and I don't believe env vars should be namespaced. I want to able to write my code as normal and expect that env vars will be available as they normally would when on a given host.

**Comment by hadim** [comment]
I agreee with @thomasdavis .

**Comment by ilkka** [comment]
@NoumanSaleem gives good advice but since envsubst isn't available on OS X without installing gettext from e.g. homebrew which apparently can cause some problems, I slapped together a Go equivalent: https://github.com/ilkka/substenv

**Comment by pmoust** [comment]
Agreed with @thomasdavis 

+1 for ENV passed on runtime

**Comment by Vad1mo** [comment]
I hate +1 but this is the most requested feature and discussed feature here.

Can we sum up the discussion here?

1. What variables are already supported in RC2 and where is it documented?
2. What is the current status?
3. Is this feature request accepted and when can one expect a solution?

**Comment by wololock** [comment]
Agreed with @Vad1mo, +1

**Comment by Vad1mo** [comment]
This is a simple script that can be used before docker-compose

https://gist.github.com/Vad1mo/9ab63f28239515d4dafd

here is the one liner to install:
`sudo curl https://gist.githubusercontent.com/Vad1mo/9ab63f28239515d4dafd/raw/aa54b91f4c3671097789a54a9f42ba679b89dbaf/replace-var -o /usr/local/bin/replace-var && sudo chmod 755 /usr/local/bin/replace-var 
`

**Comment by fullergalway** [comment]
+1

**Comment by jonathanepollack** [comment]
I hate to +1 this -- but this really is the most requested feature in github right now...

**Comment by fr3nd** [comment]
definitely +1
What's the current status? Is this feature being worked on?

**Comment by wildsurfer** [comment]
+1

**Comment by grjones** [comment]
+1

**Comment by tobijb** [comment]
+1

**Comment by coding2012** [comment]
+1, my use-case for this is to have 3 environments (a bunch of machines) that are all basically duplicates, but I want different port for each one. Rather than having a different fig file to maintain, I'd rather have a .sh file I can add a parameter like 2000, and that would spin up an environment that uses ports like 2001, 2002, 2003. I can do the addition logic in bash script, but I'll need something to pass to fig.

**Comment by noisy** [comment]
+1

**Comment by rhardih** [comment]
+1

**Comment by joshes** [comment]
+1

**Comment by montells** [comment]
+1

**Comment by pvlg** [comment]
+1

**Comment by davibe** [comment]
+1

**Comment by bpletzer** [comment]
+1

**Comment by mduheaume** [comment]
+1

**Comment by jaydeep17** [comment]
+1

**Comment by johanhaleby** [comment]
+1

**Comment by thijsterlouw** [comment]
+1

**Comment by g3rd** [comment]
+1

**Comment by hd-deman** [comment]
+1

**Comment by cpopov** [comment]
+1

**Comment by radutoev** [comment]
+1

**Comment by ggregoire** [comment]
+1

**Comment by BMouser** [comment]
+1

**Comment by christianwoehrle** [comment]
+1

**Comment by abraverm** [comment]
+1

**Comment by Vad1mo** [comment]
let me sum up the progress so far:
![image](https://cloud.githubusercontent.com/assets/1492007/7105019/4113969e-e104-11e4-83ff-c912093a3ed0.png)

**Comment by funkyfuture** [comment]
adding '+1'-comments and naughty pictures certainly don't help in a debate.

i wish i had the rights to delete them. :-P

**Comment by abraverm** [comment]
In my opinion, maintainer has limited time and energy and he needs to choose where to invest them.  '+1'  is the simplest way for a user to show which feature\bug is more important for them. There are different kind of levels of user participation in OSS projects. '+1' on the bug\feature says it in single flag what they think. And I wish you just had a button '+1' with a counter.

**Comment by javiervivanco** [comment]
+1

**Comment by docteurklein** [comment]
isn't it already possible by using the [env_file](https://docs.docker.com/compose/yml/#env_file) ?
for example: 

    echo 'VAR=test' > env && docker-compose up -d

Might not, because "Environment variables specified in environment override these values."

**Comment by spiddy** [comment]
Compose.yml file does not define just env variables. So that would not work
for setting dynamically the hostname or the exposed port number etc.

**Comment by mgasner** [comment]
+1

**Comment by justinwiley** [comment]
What are the security implications of this?  I.e. can an unprivileged user can set HOSTNAME or other environment variables referenced in the YAML to something nefarious, that would lead to unintended code execution...either through failed YAML parsing or via its use after parsing?

**Comment by spiddy** [comment]
Well, good question, I suppose if someone has privileges to execute a user-defined compose.yml file, you should assume he 'owns' the machine, since he can activate --privileged mode on a container of his choice.

If the compose.yml file is not user-defined, then the hostname can only be set if the author of the compose.yml file is letting it be defined by variable. In that case, hostname substitution shouldn't execute any arbitrary shell commands to be resolved, otherwise command injection could be possible by unprivileged users.

**Comment by jsegura** [comment]
+1

**Comment by olalonde** [comment]
+1

**Comment by dlouwers** [comment]
:+1:

**Comment by ebuildy** [comment]
Come on, even Symfony 2 has variables support in YML!

Or we need something in top of docker-compose.

**Comment by goloroden** [comment]
+1

**Comment by Ovski4** [comment]
+1

**Comment by cyppan** [comment]
+1

**Comment by bobrik** [comment]
Every time you write your `+1`, 73 people get useless email. That doesn't make any sense. I bet maintainers got your point already.

![stop](https://41.media.tumblr.com/c96ece92691dfd4fc0be3a534db30c43/tumblr_nmeqsgubgB1rn7bzro1_500.jpg)

**Comment by cyppan** [comment]
Sorry for this additional spam, but until Github supports votes in another way you're gonna see a lot of ugly, space consuming and useless +1... You have a nice car btw :p

**Comment by funkyfuture** [comment]
fyi, yesterday i requested automated censorship for '+1'-~spamming~ content.

if you want something implemented, make a sponsoring offer!

Am 30.04.2015 um 17:00 schrieb Cyprien Pannier:

**Comment by aanand** [comment]
OK! Sincere thanks for the +1s folks, it's genuinely helpful to know how many people want this feature. I've created a new issue at https://github.com/docker/compose/issues/1377 to track it. Head over there with your thoughts.

**Comment by olalonde** [comment]
<meta>In case anyone doesn't know, you can unsubscribe from notifications via the Unsubscribe button on the right.</meta>

**Comment by phazei** [comment]
I wanted to share a rather simple temporary solution I had for variable expansion:

```bash
#!/bin/sh
apply_shell_expansion() {
    declare file="$1"
    declare data=$(< "$file")
    declare delimiter="__apply_shell_expansion_delimiter__"
    declare command="cat <<$delimiter"$'\n'"$data"$'\n'"$delimiter"
    eval "$command"
}
#in case you want to store the variables in a local file
source ./local.env
if ! [ -n "$DOCKER_DATA_PATH" ] ; then
	echo "DOCKER_DATA_PATH not set, using default value"
fi
DOCKER_DATA_PATH=${DOCKER_DATA_PATH:=//c/projects/dockerdata}

apply_shell_expansion docker-compose.tpl.yml > docker-compose.yml
```

Then your `docker-compose.tpl.yml` can be commited, each person can have local.env that's sourced into the script, and it just uses bash's own variable expansion, quick and easy.

**Comment by stabenfeldt** [comment]
+1

**Comment by garystafford** [comment]
For a small number of variables ('tokens'), I use a simple shell script along with a templated version of my  yaml file. Here's an actual example:

```sh
base_url_token="{{ base_url }}" # find all these...
base_url="api.acme.com" # replace with url of public rest api
host_ip_token="{{ host_ip }}" # find all these...
host_ip=$(docker-machine ip $(docker-machine active)) # replace with ip of host running NGINX

echo ${base_url_token} = ${base_url}
echo ${host_ip_token} = ${host_ip}

sed -e "s/$base_url_token/$base_url/g" \
    -e "s/$host_ip_token/$host_ip/g" \
    < docker-compose-template.yml \
    > docker-compose.yml
```

this:
```yaml
  extra_hosts:
   - "{{ base_url }}:{{ host_ip }}"
```
becomes:
```yaml
  extra_hosts:
   - "api.acme.com:192.168.99.100"
```

**Comment by apobbati** [comment]
Github needs to implement an upvote/importance feature for issues.

**Comment by aanand** [comment]
@apobbati +1

(Environment variable interpolation has been implemented and merged in https://github.com/docker/compose/pull/1765, if that's what you're getting at.)

**Comment by funkyfuture** [comment]
<a href="https://imgflip.com/i/q3ks7"><img src="https://i.imgflip.com/q3ks7.jpg" title="made at imgflip.com"/></a>

i admit there's room for misinterpretations, so refresh your knowledge about bolshevik strategies!

**Comment by levivm** [comment]
+1
